executor/builder.go

```diff
@@ -893,25 +893,24 @@ func (b *executorBuilder) buildAnalyze(v *plan.Analyze) Executor {
 	return e
 }
 
-func (b *executorBuilder) constructDAGReq(plans []plan.PhysicalPlan) *tipb.DAGRequest {
+func constructDAGReq(plans []plan.PhysicalPlan, startTS uint64, ctx context.Context) (*tipb.DAGRequest, error) {
 	dagReq := &tipb.DAGRequest{}
-	dagReq.StartTs = b.getStartTS()
-	dagReq.TimeZoneOffset = timeZoneOffset(b.ctx)
-	sc := b.ctx.GetSessionVars().StmtCtx
+	dagReq.StartTs = startTS
+	dagReq.TimeZoneOffset = timeZoneOffset(ctx)
+	sc := ctx.GetSessionVars().StmtCtx
 	dagReq.Flags = statementContextToFlags(sc)
 	for _, p := range plans {
-		execPB, err := p.ToPB(b.ctx)
+		execPB, err := p.ToPB(ctx)
 		if err != nil {
-			b.err = errors.Trace(err)
-			return nil
+			return nil, errors.Trace(err)
 		}
 		dagReq.Executors = append(dagReq.Executors, execPB)
 	}
-	return dagReq
+	return dagReq, nil
 }
 
-func (b *executorBuilder) constructTableRanges(ts *plan.PhysicalTableScan) (newRanges []types.IntColumnRange) {
-	sc := b.ctx.GetSessionVars().StmtCtx
+func constructTableRanges(ts *plan.PhysicalTableScan, ctx context.Context) (newRanges []types.IntColumnRange, err error) {
```

Comment:
This is only used in executor builder, should be a method of it.

Reply:
No, to be used, it must be a function, `buildTableReader` function need it.
If it become a method of executor builder, `builder.err` field will be modified and DATA RACE raise.

Reply:
It doesn't have to be a function, we can still return the error instead of modifying builder.err.